v2/internal/frontend/desktop/linux/window.go

```diff
@@ -748,7 +722,11 @@ func (w *Window) ExecJS(js string) {
 }
 
 func (w *Window) StartDrag() {
-	C.StartDrag(w.webview, w.asGTKWindow())
+	data := C.DragOptions{
+		webview:    w.webview,
+		mainwindow: w.gtkWindow,
+	}
+	C.ExecuteOnMainThread(C.startDrag, C.gpointer(&data))
```

Comment:
The reason this was being constructed in C was to ensure that the DragOptions were still valid on the main thread. I'm not 100% sure on when Go will free this memory and I'm curious how this fixes the original issue!

Reply:
It's a bit weird, but the original error (end of this comment) complains about `gtk_widget_get_toplevel` not being able to assert that the given widget is a `GtkWidget`, usually. Sometimes it complained about a null pointer instead.

I tried all kinds of things to try and verify that the path from the "drag" message coming through `processMessage` had good data and valid types, but still it would fail on click.

In the end I used the simpler setup that is used by `SetRGBA` that seems to have variables with consistent access to the expected webview and mainwindow references.

For me, letting Go control and garbage collect the struct with references to the window and webview makes sense and is simpler for me to understand, but this is not an area I have a lot of experience in.

```
ian@ians-apollo:~/Projects/wails-v2-test $ build/bin/wails-v2-test
Overriding existing handler for signal 10. Set JSC_SIGNAL_FOR_GC if you want WebKit to use a different signal

(wails-v2-test:31285): Gtk-CRITICAL **: 22:20:24.689: gtk_widget_get_toplevel: assertion 'GTK_IS_WIDGET (widget)' failed

(wails-v2-test:31285): Gtk-CRITICAL **: 22:20:24.689: gtk_widget_get_toplevel: assertion 'GTK_IS_WIDGET (widget)' failed
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x7f18ae662 pc=0x7f333f293e96]

runtime stack:
runtime.throw({0x6733dc, 0x7f333af668af})
        /nix/store/76qbr3hy9adncz7j149wx9s34s0namyw-go-1.17.6/share/go/src/runtime/panic.go:1198 +0x71
runtime.sigpanic()
        /nix/store/76qbr3hy9adncz7j149wx9s34s0namyw-go-1.17.6/share/go/src/runtime/signal_unix.go:719 +0x396

goroutine 1 [syscall]:
runtime.cgocall(0x5ff1b0, 0xc00019fe80)
        /nix/store/76qbr3hy9adncz7j149wx9s34s0namyw-go-1.17.6/share/go/src/runtime/cgocall.go:156 +0x5c fp=0xc00019fe58 sp=0xc00019fe20 pc=0x40b0dc
github.com/wailsapp/wails/v2/internal/frontend/desktop/linux._Cfunc_gtk_main()
        _cgo_gotypes.go:979 +0x45 fp=0xc00019fe80 sp=0xc00019fe58 pc=0x5ecf65
github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.(*Window).Run(0xc000192230)
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/window.go:718 +0x8c fp=0xc00019fe98 sp=0xc00019fe80 pc=0x5f506c
github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.(*Frontend).Run(0xc0001921c0, {0x6da120, 0xc00007ff80})
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/frontend.go:130 +0xbd fp=0xc00019fed8 sp=0xc00019fe98 pc=0x5ef59d
github.com/wailsapp/wails/v2/internal/appng.(*App).Run(0xc00007c320)
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/appng/app_production.go:37 +0x31 fp=0xc00019ff10 sp=0xc00019fed8 pc=0x5fced1
github.com/wailsapp/wails/v2.Run(0xc0000001a0)
        /home/ian/Projects/github.com/wailsapp/wails/v2/wails.go:24 +0x33 fp=0xc00019ff28 sp=0xc00019ff10 pc=0x5fd633
main.main()
        /home/ian/Projects/wails-v2-test/main.go:26 +0x3af fp=0xc00019ff80 sp=0xc00019ff28 pc=0x5fdaaf
runtime.main()
        /nix/store/76qbr3hy9adncz7j149wx9s34s0namyw-go-1.17.6/share/go/src/runtime/proc.go:255 +0x227 fp=0xc00019ffe0 sp=0xc00019ff80 pc=0x43d0e7
runtime.goexit()
        /nix/store/76qbr3hy9adncz7j149wx9s34s0namyw-go-1.17.6/share/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc00019ffe8 sp=0xc00019ffe0 pc=0x469c81

goroutine 6 [chan receive]:
github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.(*Frontend).startMessageProcessor(0x0)
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/frontend.go:111 +0x59
created by github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.NewFrontend
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/frontend.go:96 +0x367

goroutine 7 [chan receive]:
github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.(*Frontend).startRequestProcessor(0x0)
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/frontend.go:280 +0x5a
created by github.com/wailsapp/wails/v2/internal/frontend/desktop/linux.NewFrontend
        /home/ian/Projects/github.com/wailsapp/wails/v2/internal/frontend/desktop/linux/frontend.go:97 +0x3a9
ian@ians-apollo:~/Projects/wails-v2-test $
```

Reply:
> The reason this was being constructed in C was to ensure that the DragOptions were still valid on the main thread. I'm not 100% sure on when Go will free this memory and I'm curious how this fixes the original issue!

IIRC putting the data on the Go-Stack might be problematic, it's sort of a similar problem which was in the go-webview2 code. CGO only makes sure this data remains there and doesn't get moved as long as the CGO call is running. After the CGO call returns the memory will get reused, especially after the Go method which called CGO has been returned. If the Main-Thread callback is getting called after this has happened, there might be other data stored at that pointer.